Resurrect windows build - rewrite xtensa-build-zephyr.sh to Python

aborisovich opened the pull request:
Four commits:

1. Fixed warning in sof-logger issued from compilers newer than GCC 9.1 "char-subscripts" by adding (int) cast to isspace functions.
2. To enable Windows build, added additional include path to sof logger so required system standard POSIX headers could be found in MSYS2 environment. Unfortunately there are macros redefinitions between standard GCC headers and src\include\sof\compiler_attributes.h. Could not find the way around it, also GCC does not provide any policy to silence this particular warning, so had add #ifndef guards in the compiler_attributes.h header.
3. Added \#ifndef guards to macros to avoid redefinitions with GCC standard headers on MSYS2.
4. Added xtensa-build-zephyr.py script that allows to build SOF with Zephyr on both Windows and Linux.

Signed-off-by: Andrey Borisovich andrey.borisovich@intel.com

lgirdwood reviewed (commented):
@aborisovich I'm guessing the shell script is not working on Windows ?
These scripts are developer aids only (e.g. should not really be used as part of build infra) so the only issue here is keeping this aligned with shell version or deprecate the shell version ? @marc-hb any comments.

aborisovich commented:
@lgirdwood yes, unfortunately xtensa-build-zephyr.sh does not work on Windows thus I made this counterpart.
Moreover it has some new features comparing to the old one:

- New flag -u that allows to clone any fork by providing URL in --clone mode
- Replaced "make" build commands with cmake --build what allows to use new CMake generators eg. Ninja

In my opinion this script should be thoroughly tested by end-users and when proved working replace existing xtensa-build-zephyr.sh. I did not remove it in this commit (xtensa-build-zephyr.sh) as I did not know that it is not used in processes or VMs.

aborisovich commented:
Everyone are welcome to test this script against bugs since I surely did not find all flaws.
Especially on linux OS where it had run only several times (since I wrote it on Windows).

lgirdwood commented:
> * New flag -u that allows to clone any fork by providing URL in --clone mode
> * Replaced "make" build commands with cmake --build what allows to use new CMake generators eg. Ninja

Best to put this in the commit message so it's obvious. 
Fwiw, I think @marc-hb is using Ninja so should be able to comment.

kv2019i reviewed (changes requested):
Thanks @aborisovich ! This looks very nice. I think we might want to keep both scripts in parallel for a while to get some mileage in various Linux systems. I do have a few minor questions on the earlier two patches, please see inline.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,370 @@
+		input_elf_file = pathlib.Path(WEST_TOP, platform_build_dir, "zephyr", "zephyr.elf")
+		# Extract metadata
+		subprocess.run(f"{smex_executable} -l {fw_ldc_file} {input_elf_file}", check=True,
+			shell=True, timeout=120)
```
shell=False
```suggestion
subprocess.run(smex_executable, "-l", fw_ldc_file, input_elf_file, check=True)
```

marc-hb reviewed (changes requested):
Very incomplete review sorry, not enough time right now. Will get back to it soon.

Thanks for this rewrite, impressive effort! Yes all Zephyr users should be cherry-picking this (`git fetch origin pull/5299/head && git cherry-pick FETCH_HEAD`) and start using it ASAP - even before it gets merged.

An important, future and post-merge test will be to switch CI to it but many people in our CI team are out this week for Chinese New Year. Later after some transition period we can delete the shell script.


One important feature missing is logging. This is very important not just to help resolve bugs in this script but even more importantly to quickly assess what went wrong in its environment (there is always something going wrong somewhere). The log level should always be increased in CI.

With the shell script you could quickly insert a `set -x` and get some logging anywhere. The script already has some `set -x` in some critical places.

A hopefully quick way to get a lot of useful logs with minimal code changes is to add a simple wrapper for all calls to `subprocess.run()`. That wrapper uses `logging.info()` to print all its arguments and then passes them all through to `subprocess.run()`. This is one of the things `west -v` does BTW.

Tested example:

```python
def run_command(*args, **kwargs):
    logging.warning("Running " + " ".join(args[0]))
    return subprocess.run(*args, **kwargs)   
```

aborisovich commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,370 @@
+	for path in paths:
+		if path.is_dir():
+			result = subprocess.run("west topdir", shell=True, timeout=5, capture_output=True,
+				check=False, cwd=path)
```
Can you please provide some reference on the capture_output deadlock problem? I could not find any.

marc-hb replied:
This will not happen as long as `west topdir` output volume is small - smaller than the buffer size of the PIPE created by `subprocess`. However once that (unknown) line is crossed then deadlock is guaranteed: `west topdir` is blocked waiting for the pipe to have more space. Meanwhile, this script is blocked waiting for the `west topdir` exit code.

https://docs.python.org/3/library/subprocess.html#popen-objects
https://www.google.com/search?q=subprocess+communicate+deadlock

marc-hb replied:
> To discard the output use stdout=DEVNULL and stderr=DEVNULL

Actually: use `stdout=DEVNULL` only let and stderr pass through (the default). Never discard stderr...

aborisovich replied:
Well, I think that you may be wrong here.
First, in the[ subprocess.run](https://docs.python.org/3/library/subprocess.html#subprocess.run) documentation they stated the following: 
> Note The newlines attribute of the file objects Popen.stdin, Popen.stdout and Popen.stderr are not updated by the Popen.communicate() method.

The link you send me https://docs.python.org/3/library/subprocess.html#popen-objects has that deadlock warning under Popen.wait() function description.
The first googling result I find is [this thread](https://stackoverflow.com/questions/14615462/why-does-communicate-deadlock-when-used-with-multiple-popen-subprocesses) seems to be related to the issue that occurred in older python versions. And this [fix](http://bugs.python.org/issue12786) seems to be the resolution but I am unsure.
But whatever it does not matter really, I will devnull them.

aborisovich commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,370 @@
+			"security-policy-settings/create-symbolic-links")
+		raise
+
+def west_init_update():
```
Inserted as python docs.

aborisovich replied:
Sorry, deleted by accident during squashing. Restored.

lgirdwood commented:
@aborisovich I will rerun CI this afternoon, looks like a server error for the container and some DUTs were not available. QB issue looks like a false positive.

lgirdwood commented:
@marc-hb @greg-intel this PR has failed Zephyr/Fuzzer build twice on what looks like container network issues. Should be completely unrelated to this PR. Other PRs have been building the container OK today. Can you take a quick check and rerun CI if you think this is just bad luck here. Thanks

aborisovich commented:
New fixes:
- removed previous policy to turn off -Werror on windows hosts. Fixed warnings by adding ifndef guards to macros in compiler_definitions.h
- @marc-hb added new flag "-vv" (more verbose) that implicitly implements -v (verbose) and additionally prints all commands invoked by subprocess.run(). Added a wrapper function as you suggested.
When -vv flag set it prints:
Running command: *command* in *directory*

marc-hb commented:
> @marc-hb @greg-intel this PR has failed Zephyr/Fuzzer build twice on what looks like container network issues. 

Always share URLs to test failures as Github does not support force pushes (it only tolerates them)

fredoh9 commented:
I am able to see the errors. Looks temporary network glitch.
https://github.com/thesofproject/sof/runs/5035230313?check_suite_focus=true

```
Err:1 http://archive.ubuntu.com/ubuntu focal/main amd64 libsigsegv2 amd64 2.12-2
  Could not connect to archive.ubuntu.com:80 (91.189.88.152), connection timed out Could not connect to archive.ubuntu.com:80 (91.189.88.142), connection timed out
```

Coincidentally we had sof docker image update in yesterday, I don't think this is direct reason for network failure, but I will take a look for any suspicious warning or error.

fredoh9 commented on src/include/sof/compiler_attributes.h:
```diff
@@ -1,21 +1,29 @@
 /* SPDX-License-Identifier: BSD-3-Clause
  *
- * Copyright(c) 2019 Intel Corporation. All rights reserved.
+ * Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
```
we don't update copyright year

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+else:
+	raise RuntimeError(f"Your operating system: {platform.system()} is not supported")
+
+platform_list = [
```
Please keep the distinction between default (`-a`) platforms and the superset of supported platforms. It's not used right now but it's been useful in the past and it's a small code change, just two variables instead of one.

aborisovich replied:
@marc-hb please elaborate I do not understand what do you mean.
Currently `-a` adds all platform names from `platform_list` to build queue. I though we support all platforms that are on that list?

marc-hb replied:
There are times when a platform works for only certain people who have the "special sauce" required: unusual dependencies, bleeding edge toolchain, temporary workarounds, etc. There has also been times when we temporarily "disabled" some troublesome platform. That's why the shell script has two tiers:
- The `-a` default platforms that everyone including CI can and should build whenever they make a change.
- The other, "in progress" or "more difficult" additional platforms.

aborisovich replied:
If I understand you correctly, you mean to change just the naming - from "all_platforms" to "supported_platforms" correct? OK done.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+		WEST_TOP=pathlib.Path(args.west_path)
+
+	if args.more_verbose:
+		args.verbose_build = True
```
Use an `args.verbose` log level instead, see comment below.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+		arg_list = ["-D" + arg for arg in arg_list]
+		args.cmake_args = " ".join(arg_list)
+
+def execute_command(command_args, stdin=None, input=None, stdout=None, stderr=None,
```
```suggestion
def execute_command(*cmd, **kwargs)
```

aborisovich replied:
No thank you, this assumes many things that may not be true.
As I understand this approach, you simply pass a list of all arguments in *cmd list.
Meanwhile, when you use subprocess.run you can pass selective arguments by name without maintaining the order or arguments.
Even subprocess run signature:
`def run(*popenargs,
        input=None, capture_output=False, timeout=None, check=False, **kwargs):`
does not throw all args into single array. And then *popenargs is passed to Popen ctor to resolve what is where.
I need to log cwd of the command, why would assume its position in the *cmd arguments list?

marc-hb replied:
`subprocess.run()` and `.popen()` have a single positional argument (which can be either a string a or a list). Maybe this would work with `def execute_command(cmd, **kwargs)`, I don't know.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+	command executed when 'more verbose' verbosity level is set."""
+	if args.more_verbose:
+		print_args = " ".join(command_args) if isinstance(command_args, list) else command_args
+		print_cwd = f"in: {cwd}" if cwd else f"in: {os.getcwd()}"
```
`kwargs['cwd']`

aborisovich replied:
Sorry @marc-hb i just don't get what you mean, looks like I still miss some skill in python. Can we skip this part?
I've rewritten subprocess.run signature in execute_command so you can:
execute_command(args, check=True, cwd=someDIr) == execute_command(args, cwd=someDir, check=True).
I think your approach negates that...

marc-hb replied:
OK let's leave this for later.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+	audio_modules_dir = pathlib.Path(WEST_TOP, "modules", "audio")
+	audio_modules_dir.mkdir(mode=511, parents=True, exist_ok=True)
+	sof_symlink = pathlib.Path(audio_modules_dir, "sof")
+	# Symlinks creation requires administrative privileges in Windows or special user rights
```
Or you could just drop the symlink feature on Windows? It should be optional, the older shell version of this was designed to support a variety of setups, users were allowed to organize their directories as they wished.

I'm worried symlinks on Windows will confuse users and generate "support requests", what do you think?

aborisovich replied:
Yes of course symlinks are not something commonly seen on Windows. It even surprised me that there is something like this. Running python with administrator rights is a pain and adding user a security policy violation is not a great approach either. I just did not find any better alternative to this. Simple "shortcut" in Windows terms won't do the job and we need to have sof as module in zephyr. So...
Well, we could clone zephyrproject using west "outside of" cloned sof repo but this seems dangerous too.
If I just assume that I can clone zephyr to your SOF_REPO/../../../../ this would be very bold assumption, we can't do that.
I just can't think of any better and easy solution for now and we need that script to start working on Windows.
Maybe lets leave it for now as is? Or maybe you have some alternatives in mind?

marc-hb replied:
You know Windows, I don't know it much. We can start with this and see what other people think as long as users can clone in any way they want and use this script only for building and not for cloning.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+		" --max-count=20", check=True)
+	execute_command(f"west init -l {zephyr_dir}", check=True)
+	create_zephyr_sof_symlink()
+	execute_command("west update zephyr hal_xtensa", check=True, timeout=300, cwd=WEST_TOP)
```
Important comment from the shell script was dropped.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+	STAGING_DIR = pathlib.Path(WEST_TOP, "build-sof-staging", "sof")
+	# smex does not use 'install -D'
+	STAGING_DIR.mkdir(mode=511, parents=True, exist_ok=True)
+	for platform in args.platforms:
```
To save a ton of whitespace you could move all this to a new, `build_one_platform()` function (I always wanted to do this in the shell script).

Then the `build_platforms()` function can probably be inlined and disappear.

marc-hb replied:
We can do this later but it will cause a lot of whitespace churn in git history. No big deal.

aborisovich replied:
Lets get it stable first.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+	# smex does not use 'install -D'
+	STAGING_DIR.mkdir(mode=511, parents=True, exist_ok=True)
+	for platform in args.platforms:
+		platform_dict = next(x for x in platform_list if x["name"] == platform)
```
Using `next` to pick the first element found is a bit cryptic, gave me pause.
```suggestion
		platform_dict = [x for x in platform_list if x["name"]][0]
```

marc-hb replied:
Or much better: make `platform_list` a dict. It's already a dict in disguise.
```python
default_platforms = {
   "apl" :
         {       
		"PLAT_CONFIG": "intel_adsp_cavs15",
		"XTENSA_CORE": "X4H3I16w2D48w3a_2017_8",
		"XTENSA_TOOLS_VERSION": f"RG-2017.8{xtensa_tools_version_postfix}"
	},
...
```

Will make this easier:
```python
default_platforms['tgl-h]' = default_platforms['tgl]'
```

aborisovich replied:
Can we please skip refactoring part for now? I know there is much to polish in this script but I really can't spare more time this now. Changed platform_dict to comprehention list as suggested.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+		fw_ldc_file = pathlib.Path(STAGING_DIR, f"sof-{platform}.ldc")
+		input_elf_file = pathlib.Path(WEST_TOP, platform_build_dir, "zephyr", "zephyr.elf")
+		# Extract metadata
+		execute_command(f"{smex_executable} -l {fw_ldc_file} {input_elf_file}", check=True)
```
I don't see how these can work with `shell=False`... should be:
```suggestion
		execute_command([smex_executable, "-l", fw_ldc_file, ,input_elf_file], check=True)
```

aborisovich replied:
It works because subprocess.run accepts a list of arguments or a single string separated with spaces.
Internally they are split into the list anyway.

aborisovich replied:
Seems that it is true only for Windows OS. Rewritten all execute_command arguments to lists.

marc-hb reviewed (changes requested):
Thanks for the updates. A quick but mostly complete review this time.

aborisovich commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+						"behavior is undefined."
+						"This -p option is always required_if the real (not symbolic)"
+						"sof/ and zephyr-project/ directories are not nested in one"
+						"another.")
```
Yeah I don't like them since you have to break indentation with them or else have that whitespace indentation in the actual message.

marc-hb replied:
Just a suggestion, we can always revisit minor details like this later after merge.

aborisovich commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,382 @@
+
+# Resolve SOF_TOP directory as: "this script directory/.."
+SOF_TOP=pathlib.Path(__file__).parents[1].resolve()
+WEST_TOP=pathlib.Path(SOF_TOP, "zephyrproject")
```
Yeah I use pylint. Just wanted to have a clear similarity to bash counterpart. Ok will change to lowercase.

marc-hb replied:
To be clear I'm not sure which case is best or pylint compliant, I just don't want readers to assume this is a constant.

aborisovich commented:
Fixed most of the issues.
@marc-hb thanks especially for that tip with --verbose part.
Can we skip those refactoring parts with dictionary and execute_command signature?
I am not that confident with this arg passing part and I am afraid to break something.
Meanwhile every change to the script requires testing what takes time.
Can we please add this to the process where this script will be actually used?
Then we can refactor further and easily check if something is not broken much faster.

aborisovich commented:
Got some test failure in test-speaker.sh.
It is possible that this is false-positive?
Changes in code are so marginal... Can I re-run it somehow?

lgirdwood commented:
> Got some test failure in test-speaker.sh. It is possible that this is false-positive? Changes in code are so marginal... Can I re-run it somehow?

Its unrelated and the DUTs sometimes have false positives now and then.

kv2019i reviewed (approved):
Thanks @aborisovich ! My concerns now all addressed! I'll approve for my part but for merging you'd need to get rid of the merge commit.

lgirdwood commented:
@marc-hb good for you ?

marc-hb commented:
Let me test it today

marc-hb reviewed (changes requested):
I just tested commit 5fe03a54f8ac1 cherry-picked on top of 99b4bdffd279
```
$ ./scripts/xtensa-build-zephyr.py -v -v apl

Found west: /srv/mherber2/.local/bin/west
Building platforms: apl
```
And that's it. Nothing happens.

Then I tried a simple `-c` and it failed, see below. The `-c` option is required by our CI; it clones only once and then iterates on all platforms.

I really don't mind merging with a few bugs but these are basic features in a script that does not have that many :-(


```
--- Creating /srv/marc/SOF/sof/zephyrproject/.west and local configuration file
=== Initialized. Now run "west update" inside /srv/marc/SOF/sof/zephyrproject.
=== updating hal_xtensa (modules/hal/xtensa):
--- hal_xtensa: initializing
Initialized empty Git repository in /srv/marc/SOF/sof/zephyrproject/modules/hal/xtensa/.git/
--- hal_xtensa: fetching, need revision 0e577021bb66e644afd067cd9f7c71ab11b62b3d
remote: Enumerating objects: 294, done.
remote: Counting objects: 100% (294/294), done.
remote: Compressing objects: 100% (163/163), done.
remote: Total 294 (delta 140), reused 233 (delta 96), pack-reused 0
Receiving objects: 100% (294/294), 376.50 KiB | 2.46 MiB/s, done.
Resolving deltas: 100% (140/140), done.
From https://github.com/zephyrproject-rtos/hal_xtensa
 * [new branch]                RF         -> refs/west/RF
 * [new branch]                master     -> refs/west/master
 * [new branch]                zephyr     -> refs/west/zephyr
 * [new tag]                   RF-2015.2  -> RF-2015.2
HEAD is now at 0e577021bb66 zephyr: add SoC overlay for NXP i.MX8MP
HEAD is now at 0e577021bb66 zephyr: add SoC overlay for NXP i.MX8MP
SOF_TOP=/srv/marc/SOF/sof
west_top=/srv/marc/SOF/sof/zephyrproject
Traceback (most recent call last):
  File "./scripts/xtensa-build-zephyr.py", line 390, in <module>
    main()
  File "./scripts/xtensa-build-zephyr.py", line 387, in main
    run_clone_mode()
  File "./scripts/xtensa-build-zephyr.py", line 360, in run_clone_mode
    build_platforms()
  File "./scripts/xtensa-build-zephyr.py", line 349, in build_platforms
    sof_logger_executable_to_copy = pathlib.Path(west_top, platform_build_dir_name, "zephyr",
UnboundLocalError: local variable 'platform_build_dir_name' referenced before assignment
```

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,390 @@
+	create_zephyr_directory()
+	west_init_update()
+	git_submodules_update()
+	build_platforms()
```
`build_platforms()` belongs to `main()`, just like in the original script. It must be possible to (re)build without cloning or creating a symbolic link or any other once-off setup step.

Just copy the logic of the original script for now.

aborisovich replied:
Logical changes already started, you know if I already spend time resolving this matters why should I drop it? Its my work after all. If you wish to rebuild simply do
`./scripts/xtensa-build-zephyr.py -p <your_SOF_dir>/zephyrproject -v -v <platform>`
Maybe I will add new logic for convenience - if -p flag is empty then resolve it to <your_SOF_dir>/zephyrproject by default. The only use case with providing paths I can think of is when you would clone multiple zephyr repositories under different names into the SOF_TOP directory and would like to point to different ones lets say building platformX with zephyrproject1_dir and building platformY with zephyrproject2_dir. I assume it was original goal of the xtensa-build-zephyr.sh and leave it like this.

marc-hb replied:
I really don't see which actual problems are solved by moving the single, `build_platforms()` line out of `main()`. The only effect I've seen so far has been to cost me at least 15 min trying to understand why nothing builds. Not something we want every user to experience and it would cost a lot more in validation.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,390 @@
+	else:
+		print("Building platforms: {}".format(" ".join(args.platforms)))
+
+	# Two mutually exclusive working modes are supported
```
Cloning and linking are not mutually exclusive. The two options that are mutually exclusive are:
- Pointing at an _existing_ zephyr clone with `-p`
- Downloading a _new_ zephyr clone with `-c`

These are not two "modes", they're just two (obviously) mutually exclusive command line flags and features.

For now just copy the logic from the `main()` function in the original shell script, it's only 50 lines of commented code. It's also well tested so it barely needs testing again. You almost don't even need to understand it :-) Copy the comments too.

aborisovich replied:
@marc-hb I guess we have some misunderstanding here.
I named it "link_mode" as it "links existing zephyr directory using symlink" that's what I meant (not the linking process of linker). Logic is copied as is but with some additional features. That being said, that .sh script is not a holy grail and I can and will reword, reinterpret or rephrase content of it as a developer of this script as a part of my contribution.
Those two arguments "-c" and "-p" are not "flags". A flag provides user input to some selected program behavior. Those are "features" but not in the context of this code and review. They are exactly what i called them -" modes". Two different program logic flows. Can we please accept my point of view and understanding? I think they are perfectly valid and should not be a subject of this review. We should focus on finding bugs/errors in the script or use cases when it does not do what it supposed to do.

aborisovich replied:
Changed function name to run_point_mode() for consistency.

marc-hb replied:
> Can we please accept my point of view and understanding for now?

"flags", "modes", etc. I don't really care how you call them; afraid you lost me in the various definitions and nuances. My point was entirely the first sentence: "Cloning and linking are not mutually exclusive", I think that's why nothing built the first time I tried to use this script. I thought the "modes" terminology would help explain the problem but forget it, let's leave terminology aside.

I don't understand why you have to "fight argparser", some argument validation and processing can done _after_ invoking argparse, it does not have to be entirely implemented in argparse.

Last but not least, do not "resolve" and hide discussions after asking  "can you accept my point of view?" + leaving the "mutually exclusive" comment in place. Discussions are "resolved" only after agreement and it's fine to leave some discussions never resolved / never hidden - and easier to see even after merge.

aborisovich commented:
@marc-hb yes we need both working modes and both systems, no sense to merge if it does not work. I'll try to push fixes today.

marc-hb reviewed (commented):
> If you wish to rebuild simply do ...

The initial goal is to replace the shell script so everyone is using the same script ASAP. So in the very first version of this Python script the most basic build interface must work the same so it does not break anyone's build and does not force them to investigate what goes wrong. People have more urgent and important things to do.

As in earlier https://github.com/thesofproject/rimage/pull/87#discussion_r792329665 you seem to have little consideration for breaking the build for existing users. This is not acceptable, please stop. Disrupting other users and validation is of course necessary sometimes but there needs to be some value in return and so far you have not pointed at any such value in this PR.


> Logical changes already started, you know if I already spend time resolving this matters why should I drop it? Its my work after all.

Resolving which problems? If you have some additional fixes compared to the shell script then great but why should they change the most basic user interface? It's not a very complex user interface.

Building Zephyr is something that a number of developers and validation already do many times every day. If "your work after all" must interfere with that for some reason then let's discuss it later in follow-up, smaller and more focused PRs once everyone has switched and when we can coordinate any backwards incompatible changes across all users and validation. 

> I assume it was original goal of the xtensa-build-zephyr.sh and leave it like this.

The fastest is really to not guess or assume anything and start by reproducing the simple logic of the initial script. If not then why not?

aborisovich commented:
@marc-hb Resolved issues:

-  `./scripts/xtensa-build-zephyr.py -c` tried to build platforms when no platforms provided. Fixed - when no platforms provided it just clones and updates submodules as it should.
- `./scripts/xtensa-build-zephyr.py -v -v apl` did nothing as no -p flag was specified. Fixed - now when -p or -c flag specified, -p with default directory SOF_TOP/zephyrproject is used.
- fixed issue with XTENSA_TOOLS_ROOT variable not defined in the environment. Fixed - now if platform selected for build has XTENSA_TOOLS_VERSION in its dictionary, XTENSA_TOOLS_ROOT is read from environment or exception with informative message thrown.
- help description of `--all` flag had been reworded so it is clear that it builds all **currently supported** platforms.

marc-hb commented:
Thanks, I will test again in the next 24h

aborisovich commented:
> "flags", "modes", etc. I don't really care how you call them; afraid you lost me in the various definitions and nuances. My point was entirely the first sentence: "Cloning and linking are not mutually exclusive", I think that's why nothing built the first time I tried to use this script. I thought the "modes" terminology would help explain the problem but forget it, let's leave terminology aside.

I don't understand why you have to "fight argparser", some argument validation and processing can done after invoking argparse, it does not have to be entirely implemented in argparse.

Last but not least, do not "resolve" and hide discussions after asking "can you accept my point of view?" + leaving the "mutually exclusive" comment in place. Discussions are "resolved" only after agreement and it's fine to leave some discussions never resolved / never hidden - and easier to see even after merge.

Ok sorry I will not resolve them. I just did not get your point fully and I am also getting confused in all this. It is a complex script no doubt. I also got carried away for a while with that argparser subparser idea. For some reason I imagined for a while that we need both -p argument signaling that we actually want to rebuild and another -P argument giving a path to zephyr project.
Fortunately I realized it is not needed at all and it should be used just as you indicated - if no -p or -c flag specified just do -p with default path.

marc-hb commented:
I don't mind merging even with some bugs or room for improvement. I mind merging with a different user interface / "API" (at least not without some strong reason) because bringing everyone on the same page again will disrupt at least half the users.

I will give this another quick try today as promised.

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,399 @@
+		if "XTENSA_TOOLS_VERSION" in platform_dict:
+			XTENSA_TOOLS_ROOT = os.getenv("XTENSA_TOOLS_ROOT")
+			if not XTENSA_TOOLS_ROOT:
+				raise RuntimeError(f"Platform {platform} uses Xtensa toolchain."
```
This makes it impossible to build with the Zephyr SDK.

So far we've been able to build every platform with either the Zephyr SDK or XCC, both work fine. The Zephyr SDK requires zero license and zero configuration, install it in the HOME directory and that's it, job done. Nothing to do in this script to support it (besides removing this RunTimeError)

marc-hb reviewed (changes requested):
Thanks for the cloning fixes.

Tried it again, now clones but fails to build with:
```
    raise RuntimeError(f"Platform {platform} uses Xtensa toolchain."
RuntimeError: Platform apl uses Xtensa toolchain.
Please set your XTENSA_TOOLS_ROOT environmental variable
```
There is no platform that requires XCC.

Again this would have not happened if you had simply copied the original logic.

We're making progress but using me as CI is really not efficient. Try this tiny test hack instead and github will test the most basic features (clone and build) for you:

```diff
diff --git a/zephyr/docker-build.sh b/zephyr/docker-build.sh
index 0700cbaa780b..d04ea21be028 100755
--- a/zephyr/docker-build.sh
+++ b/zephyr/docker-build.sh
@@ -36,5 +36,5 @@ else
     stat .
     sudo chgrp -R 1000 .
     sudo chmod -R g+rwX .
-    ./scripts/xtensa-build-zephyr.sh -a -c "$@"
+    ./scripts/xtensa-build-zephyr.py -a -c "$@"
 fi
```

aborisovich commented:
@marc-hb fixed exception issue with XTENSA_TOOLS_ROOT - now when XTENSA_TOOLS_ROOT environmental variable is set it is a signal to use user installed custom Xtensa installation. When not assume user has Zephyr SDK installed.
Performed manual tests on linux:

1. Command `./scripts/xtensa-build-zephyr.py -c -a` works now exactly the same as  `./scripts/xtensa-build-zephyr.sh -c -a`. All supported platforms are built successfully using Zephyr SDK.
2. Command `./scripts/xtensa-build-zephyr.py -c tgl tgl-h` with environmental variable: XTENSA_TOOLS_ROOT set works exactly the same as `./scripts/xtensa-build-zephyr.sh -c tgl tgl-h`. User custom xtensa installation is used and platforms are successfully built. Unfortunately could not test -a flag as I could not find Xtensa cores to download for APL. I think they are unavailable now.
3. Command `./scripts/xtensa-build-zephyr.py -c tgl imx8m` with environmental variable: XTENSA_TOOLS_ROOT set works exactly the same as `./scripts/xtensa-build-zephyr.sh -c tgl imx8m` - does not work because scripts are unable to 'deinitialize' xtensa toolchain when building imx8m that was previously initialized to build tgl. But well maybe we will improve it someday.

Added new features:

- Added pretty printing to similar to linux `tree` command that xtensa-build-zephyr.sh uses to present output directory contents.

Created test [PR](https://github.com/thesofproject/sof/pull/5353) replacing invocation of the xtensa-build-zephyr.sh with new script. Build Zephyr workflow is passing. Currently Windows build encountered regression as I am failing to build sof sources due to zephyr libc clashes when using standard gcc. I would submit this script for now without replacing existing xtensa-build-zephyr.sh. Passing CI build does not mean this script is fully mature and tested taking into consideration its complexity...

paulstelian97 commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,432 @@
+	{
+		"name": "imx8",
+		"PLAT_CONFIG": "nxp_adsp_imx8",
+		"RIMAGE_KEY": "" # no key needed for imx8
```
Why the inconsistency between empty/not needed vs ignored between the NXP platforms? There's nothing intrinsically different between imx8 and imx8x platforms.

Is this inconsistency in the original script?

paulstelian97 replied:
Guess that's fine then, I'd have to come and fix it later myself then.

aborisovich commented:
Minor update - installed files are not presented when platforms were not built (no need to do that when somebody uses only -c flag).

marc-hb commented on scripts/xtensa-build-zephyr.py:
```diff
@@ -0,0 +1,432 @@
+	elif not args.clone_mode:	# if neather -p nor -c provided, use -p
+		args.west_path = west_top
+
+	if args.verbose == 1:
```
Maybe ?
```suggestion
	if args.verbose >= 1:
```

marc-hb reviewed (approved):
All basic features work for me, thanks @aborisovich very impressive effort.

I will help start the transition away from the .sh and we can refine if needed.

lgirdwood reviewed (approved):
Good work @aborisovich !
